api_app/service_bus/airlock_request_status_update.py

```diff
@@ -0,0 +1,98 @@
+from db.repositories.airlock_requests import AirlockRequestRepository
+from models.domain.airlock_operations import StepResultStatusUpdateMessage
+from service_bus.helpers import default_credentials
+from core import config
+from db.errors import EntityDoesNotExist
+from resources import strings
+
+
+async def receive_message_from_step_result_queue():
+    """
+    This method is an async generator which receives messages from service bus
+    and yields those messages. If the yielded function return True the message is
+    marked complete.
+    """
+    async with default_credentials() as credential:
+        service_bus_client = ServiceBusClient(config.SERVICE_BUS_FULLY_QUALIFIED_NAMESPACE, credential)
+
+        async with service_bus_client:
+            receiver = service_bus_client.get_queue_receiver(queue_name=config.SERVICE_BUS_STEP_RESULT_QUEUE)
+
+            async with receiver:
+                received_msgs = await receiver.receive_messages(max_message_count=10, max_wait_time=5)
+
+                for msg in received_msgs:
+                    result = True
+                    message = ""
+
+                    try:
+                        message = json.loads(str(msg))
+                        result = (yield parse_obj_as(StepResultStatusUpdateMessage, message))
+                    except (json.JSONDecodeError, ValidationError) as e:
+                        logging.error(f"{strings.STEP_RESULT_MESSAGE_FORMAT_INCORRECT}: {e}")
+
+                    if result:
+                        logging.info(f"Received step_result status update message with correlation ID {msg.correlation_id}: {message}")
+                        await receiver.complete_message(msg)
+
+
+async def update_status_in_database(airlock_request_repo: AirlockRequestRepository, step_result_message: StepResultStatusUpdateMessage):
+    """
+    If there is another step in the operation after this one, process the substitutions + patch, then enqueue a message to process it.
+    """
+    result = False
+    try:
+        step_result_data = step_result_message.data
+        airlock_request_id = step_result_data.request_id
+        current_status = step_result_data.completed_step
+        new_status = AirlockRequestStatus(step_result_data.new_status)
+        # Find the airlock request by id
+        airlock_request = await get_airlock_request_by_id_from_path(airlock_request_id=airlock_request_id, airlock_request_repo=airlock_request_repo)
+        # Validate that the airlock request status is the same as current status
+        if airlock_request.status == current_status:
+            # update to new status and send to event grid
+            await update_status_and_publish_event_airlock_request(airlock_request=airlock_request, airlock_request_repo=airlock_request_repo, user=airlock_request.user, new_status=new_status)
+            result = True
+        else:
+            logging.error(strings.STEP_RESULT_MESSAGE_INVALID_STATUS)
```

Comment:
you can add more details to the error. request id status, new status etc.

do you need to set the result here as well to true?

Reply:
Changed